cross-compiled CUDA builds running out of disk space

I don't know what change caused this (perhaps something in the CUDA setup...), but since about a month or so, cross-compiling CUDA consistently blows through the disk space of the azure workers, failing the job in an often un-restartable way. 

I've tried fixing this in various ways (#1075, #1081, 7c267128a11af2d62855898d2812297fd9ecacc0, a8ca8f793c920dad4dfb2f377a53ab886c8603d3, 555a42cdd11c50a3c7ac5d105248e62d91982525). The problem exists on both ppc & aarch, but for ppc at least, the various fixes seem to have mostly settled things, but for aarch it's still failing 9 times out of 10.

(note, the only reason I disabled aws-sdk-cpp is that jobs started failing again with migrating to a new version that had some more features enabled, and a footprint around 40MB; this is being tackled in https://github.com/conda-forge/google-cloud-cpp-feedstock/issues/141).

CC @conda-forge/cuda-compiler @jakirkham @isuruf

PS. By chance I've seen that qt also has disk space problems, and worked around this by [disabling](https://github.com/conda-forge/qt-main-feedstock/blob/eb39f78c2e2c2c5c67c41b809813a9ea8b1dc028/recipe/build.sh#L66-L70) the use of precompiled headers. Arrow has an option `ARROW_USE_PRECOMPILED_HEADERS`, but it's already off [by default](https://github.com/apache/arrow/blob/apache-arrow-12.0.1/cpp/cmake_modules/DefineOptions.cmake#L176).

However it _does_ kinda seem related to PCH's, in that the failure looks like:
```
[73/134] Building CXX object CMakeFiles/gandiva.dir/cmake_pch.hxx.gch
FAILED: CMakeFiles/gandiva.dir/cmake_pch.hxx.gch 
$BUILD_PREFIX/bin/aarch64-conda-linux-gnu-c++ -Dgandiva_EXPORTS -I$SRC_DIR/python/pyarrow/src -I$SRC_DIR/python/build/temp.linux-aarch64-cpython-311/pyarrow/src -isystem $PREFIX/include/python3.11 -isystem /home/conda/feedstock_root/build_artifacts/apache-arrow_1688985855110/_build_env/venv/lib/python3.11/site-packages/numpy/core/include -Wno-noexcept-type  -Wall -fno-semantic-interposition -fvisibility-inlines-hidden -std=c++17 -fmessage-length=0 -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O3 -pipe -isystem $PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/pyarrow-12.0.1 -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix -isystem /usr/local/cuda/targets/sbsa-linux/include -fdiagnostics-color=always  -fno-omit-frame-pointer -Wno-unused-variable -Wno-maybe-uninitialized -O3 -DNDEBUG -O2 -ftree-vectorize -std=c++17 -fPIC -Winvalid-pch -x c++-header -include $SRC_DIR/python/build/temp.linux-aarch64-cpython-311/CMakeFiles/gandiva.dir/cmake_pch.hxx -MD -MT CMakeFiles/gandiva.dir/cmake_pch.hxx.gch -MF CMakeFiles/gandiva.dir/cmake_pch.hxx.gch.d -o CMakeFiles/gandiva.dir/cmake_pch.hxx.gch -c $SRC_DIR/python/build/temp.linux-aarch64-cpython-311/CMakeFiles/gandiva.dir/cmake_pch.hxx.cxx
In file included from $SRC_DIR/python/pyarrow/src/arrow/python/platform.h:28,
                 from $SRC_DIR/python/pyarrow/src/arrow/python/pch.h:24,
                 from $SRC_DIR/python/build/temp.linux-aarch64-cpython-311/CMakeFiles/gandiva.dir/cmake_pch.hxx:5,
                 from <command-line>:
$PREFIX/include/python3.11/datetime.h:264:1: fatal error: cannot write PCH file: No space left on device
  264 | }
      | ^
compilation terminated.
```

CC @kou @pitrou

--- a/.azure-pipelines/azure-pipelines-linux.yml
+++ b/.azure-pipelines/azure-pipelines-linux.yml
@@ -8,10 +8,10 @@ jobs:
     vmImage: ubuntu-latest
   strategy:
     matrix:
-      linux_64_cuda_compiler_version10.2:
-        CONFIG: linux_64_cuda_compiler_version10.2
+      linux_64_cuda_compiler_version11.2:
+        CONFIG: linux_64_cuda_compiler_version11.2
         UPLOAD_PACKAGES: 'True'
-        DOCKER_IMAGE: quay.io/condaforge/linux-anvil-cos7-cuda:10.2
+        DOCKER_IMAGE: quay.io/condaforge/linux-anvil-cuda:11.2
       linux_64_cuda_compiler_versionNone:
         CONFIG: linux_64_cuda_compiler_versionNone
         UPLOAD_PACKAGES: 'True'
@@ -36,10 +36,27 @@ jobs:
 
   steps:
   - script: |
-         rm -rf /opt/ghc
+         sudo mkdir -p /opt/empty_dir || true
+         for d in \
+                  /opt/ghc \
+                  /opt/hostedtoolcache \
+                  /usr/lib/jvm \
+                  /usr/local/.ghcup \
+                  /usr/local/lib/android \
+                  /usr/local/share/powershell \
+                  /usr/share/dotnet \
+                  /usr/share/swift \
+                  ; do
+           sudo rsync --stats -a --delete /opt/empty_dir/ $d || true
+         done
+         sudo apt-get purge -y -f firefox \
+                                  google-chrome-stable \
+                                  microsoft-edge-stable
+         sudo apt-get autoremove -y >& /dev/null
+         sudo apt-get autoclean -y >& /dev/null
+         sudo docker image prune --all --force
          df -h
     displayName: Manage disk space
-
   # configure qemu binfmt-misc running.  This allows us to run docker containers
   # embedded qemu-static
   - script: |

--- a/.azure-pipelines/azure-pipelines-win.yml
+++ b/.azure-pipelines/azure-pipelines-win.yml
@@ -8,8 +8,8 @@ jobs:
     vmImage: windows-2022
   strategy:
     matrix:
-      win_64_cuda_compiler_version10.2:
-        CONFIG: win_64_cuda_compiler_version10.2
+      win_64_cuda_compiler_version11.2:
+        CONFIG: win_64_cuda_compiler_version11.2
         UPLOAD_PACKAGES: 'True'
       win_64_cuda_compiler_versionNone:
         CONFIG: win_64_cuda_compiler_versionNone

--- a/.ci_support/linux_64_cuda_compiler_version11.2.yaml
+++ b/.ci_support/linux_64_cuda_compiler_version11.2.yaml
@@ -7,9 +7,9 @@ bzip2:
 c_compiler:
 - gcc
 c_compiler_version:
-- '7'
+- '10'
 cdt_name:
-- cos6
+- cos7
 channel_sources:
 - conda-forge
 channel_targets:
@@ -17,15 +17,15 @@ channel_targets:
 cuda_compiler:
 - nvcc
 cuda_compiler_version:
-- '10.2'
+- '11.2'
 cuda_compiler_version_min:
-- '10.2'
+- '11.2'
 cxx_compiler:
 - gxx
 cxx_compiler_version:
-- '7'
+- '10'
 docker_image:
-- quay.io/condaforge/linux-anvil-cos7-cuda:10.2
+- quay.io/condaforge/linux-anvil-cuda:11.2
 gflags:
 - '2.2'
 glog:

--- a/.ci_support/linux_64_cuda_compiler_versionNone.yaml
+++ b/.ci_support/linux_64_cuda_compiler_versionNone.yaml
@@ -19,7 +19,7 @@ cuda_compiler:
 cuda_compiler_version:
 - None
 cuda_compiler_version_min:
-- '10.2'
+- '11.2'
 cxx_compiler:
 - gxx
 cxx_compiler_version:

--- a/.ci_support/migrations/aws_crt_cpp0203.yaml
+++ b/.ci_support/migrations/aws_crt_cpp0203.yaml
@@ -1,8 +1,0 @@
-__migrator:
-  build_number: 1
-  kind: version
-  migration_number: 1
-  automerge: true
-aws_crt_cpp:
-- 0.20.3
-migrator_ts: 1688632499.3242948

--- a/.ci_support/migrations/libgrpc154.yaml
+++ b/.ci_support/migrations/libgrpc154.yaml
@@ -1,7 +1,0 @@
-__migrator:
-  build_number: 1
-  kind: version
-  migration_number: 1
-libgrpc:
-- '1.54'
-migrator_ts: 1681882973.1792915

--- a/.ci_support/migrations/orc190.yaml
+++ b/.ci_support/migrations/orc190.yaml
@@ -1,7 +1,0 @@
-__migrator:
-  build_number: 1
-  kind: version
-  migration_number: 1
-migrator_ts: 1688114401.8279328
-orc:
-- 1.9.0

--- a/.ci_support/win_64_cuda_compiler_version11.2.yaml
+++ b/.ci_support/win_64_cuda_compiler_version11.2.yaml
@@ -15,9 +15,9 @@ channel_targets:
 cuda_compiler:
 - nvcc
 cuda_compiler_version:
-- '10.2'
+- '11.2'
 cuda_compiler_version_min:
-- '10.2'
+- '11.2'
 cxx_compiler:
 - vs2019
 gflags:

--- a/.ci_support/win_64_cuda_compiler_versionNone.yaml
+++ b/.ci_support/win_64_cuda_compiler_versionNone.yaml
@@ -17,7 +17,7 @@ cuda_compiler:
 cuda_compiler_version:
 - None
 cuda_compiler_version_min:
-- '10.2'
+- '11.2'
 cxx_compiler:
 - vs2019
 gflags:

--- a/README.md
+++ b/README.md
@@ -67,10 +67,10 @@ Current build status
         <table>
           <thead><tr><th>Variant</th><th>Status</th></tr></thead>
           <tbody><tr>
-              <td>linux_64_cuda_compiler_version10.2</td>
+              <td>linux_64_cuda_compiler_version11.2</td>
               <td>
                 <a href="https://dev.azure.com/conda-forge/feedstock-builds/_build/latest?definitionId=54&branchName=main">
-                  <img src="https://dev.azure.com/conda-forge/feedstock-builds/_apis/build/status/arrow-cpp-feedstock?branchName=main&jobName=linux&configuration=linux%20linux_64_cuda_compiler_version10.2" alt="variant">
+                  <img src="https://dev.azure.com/conda-forge/feedstock-builds/_apis/build/status/arrow-cpp-feedstock?branchName=main&jobName=linux&configuration=linux%20linux_64_cuda_compiler_version11.2" alt="variant">
                 </a>
               </td>
             </tr><tr>
@@ -123,10 +123,10 @@ Current build status
                 </a>
               </td>
             </tr><tr>
-              <td>win_64_cuda_compiler_version10.2</td>
+              <td>win_64_cuda_compiler_version11.2</td>
               <td>
                 <a href="https://dev.azure.com/conda-forge/feedstock-builds/_build/latest?definitionId=54&branchName=main">
-                  <img src="https://dev.azure.com/conda-forge/feedstock-builds/_apis/build/status/arrow-cpp-feedstock?branchName=main&jobName=win&configuration=win%20win_64_cuda_compiler_version10.2" alt="variant">
+                  <img src="https://dev.azure.com/conda-forge/feedstock-builds/_apis/build/status/arrow-cpp-feedstock?branchName=main&jobName=win&configuration=win%20win_64_cuda_compiler_version11.2" alt="variant">
                 </a>
               </td>
             </tr><tr>

--- a/conda-forge.yml
+++ b/conda-forge.yml
@@ -1,4 +1,5 @@
 azure:
+  free_disk_space: true
   max_parallel: 20
 bot:
   abi_migration_branches:

--- a/recipe/build-arrow.sh
+++ b/recipe/build-arrow.sh
@@ -53,9 +53,6 @@ if [[ "$target_platform" == "linux-ppc64le" ]]; then
   CXXFLAGS="$(echo $CXXFLAGS | sed 's/-fno-plt //g')"
 fi
 
-# this symbol affects the ABI also on unix, see https://github.com/apache/arrow/issues/35987
-CXXFLAGS="$CXXFLAGS -DPROTOBUF_USE_DLLS"
-
 if [[ "${target_platform}" == "linux-aarch64" ]] || [[ "${target_platform}" == "linux-ppc64le" ]]; then
     # Limit number of threads used to avoid hardware oversubscription
     export CMAKE_BUILD_PARALLEL_LEVEL=3

--- a/recipe/build-arrow.sh
+++ b/recipe/build-arrow.sh
@@ -56,14 +56,9 @@ fi
 # this symbol affects the ABI also on unix, see https://github.com/apache/arrow/issues/35987
 CXXFLAGS="$CXXFLAGS -DPROTOBUF_USE_DLLS"
 
-ARROW_WITH_GCS=ON
 if [[ "${target_platform}" == "linux-aarch64" ]] || [[ "${target_platform}" == "linux-ppc64le" ]]; then
     # Limit number of threads used to avoid hardware oversubscription
     export CMAKE_BUILD_PARALLEL_LEVEL=3
-    # see meta.yaml
-    if [[ "${cuda_compiler_version}" != "None" ]]; then
-        ARROW_WITH_GCS=OFF
-    fi
 fi
 
 # reusable variable for dependencies we cannot yet unvendor
@@ -90,7 +85,7 @@ cmake -GNinja \
     -DARROW_FLIGHT_SQL=ON \
     -DARROW_GANDIVA=ON \
     -DARROW_GANDIVA_PC_CXX_FLAGS="${ARROW_GANDIVA_PC_CXX_FLAGS}" \
-    -DARROW_GCS=${ARROW_WITH_GCS} \
+    -DARROW_GCS=ON \
     -DARROW_GDB_INSTALL_DIR=replace_this_section_with_absolute_slashed_path_to_CONDA_PREFIX/lib \
     -DARROW_HDFS=ON \
     -DARROW_JEMALLOC=ON \

--- a/recipe/build-pyarrow.sh
+++ b/recipe/build-pyarrow.sh
@@ -41,10 +41,6 @@ fi
 if [[ "${target_platform}" == "linux-aarch64" ]] || [[ "${target_platform}" == "linux-ppc64le" ]]; then
     # Limit number of threads used to avoid hardware oversubscription
     export CMAKE_BUILD_PARALLEL_LEVEL=4
-    # see meta.yaml
-    if [[ "${cuda_compiler_version}" != "None" ]]; then
-        PYARROW_WITH_GCS=0
-    fi
 fi
 
 cd python

--- a/recipe/conda_build_config.yaml
+++ b/recipe/conda_build_config.yaml
@@ -1,3 +1,5 @@
+# keep this in sync with llvm_version in meta.yaml;
+# don't rely on global pinning even if it matches
 c_compiler_version:            # [osx]
   - 15                         # [osx]
 cxx_compiler_version:          # [osx]

--- a/recipe/meta.yaml
+++ b/recipe/meta.yaml
@@ -111,9 +111,7 @@ outputs:
         # - proto-opentelemetry-proto =={{ cpp_opentelemetry_sdk }}
         - gflags
         - glog
-        # constantly runs out of disk space in CI; for now either gcp- OR cuda-support on aarch/ppc
-        # retry again after https://github.com/conda-forge/google-cloud-cpp-feedstock/issues/141
-        - google-cloud-cpp  # [not ((aarch64 or ppc64le) and cuda_compiler_version != "None")]
+        - google-cloud-cpp
         # arrow uses a customized jemalloc, see #944
         # - jemalloc
         - libabseil

--- a/recipe/meta.yaml
+++ b/recipe/meta.yaml
@@ -21,7 +21,7 @@ source:
     folder: testing
 
 build:
-  number: 5
+  number: 6
   # for cuda support, building with one version is enough to be compatible with
   # all later versions, since arrow is only using libcuda, and not libcudart.
   skip: true  # [cuda_compiler_version not in ("None", cuda_compiler_version_min)]

--- a/recipe/meta.yaml
+++ b/recipe/meta.yaml
@@ -59,7 +59,7 @@ outputs:
         - exit 0
 
   - name: libarrow
-    script: build-arrow.sh  # [unix]
+    script: build-arrow.sh   # [unix]
     script: build-arrow.bat  # [win]
     version: {{ version }}
     build:
@@ -211,7 +211,7 @@ outputs:
         - exit 0
 
   - name: pyarrow
-    script: build-pyarrow.sh  # [unix]
+    script: build-pyarrow.sh   # [unix]
     script: build-pyarrow.bat  # [win]
     version: {{ version }}
     build:
@@ -302,7 +302,7 @@ outputs:
       summary: Python libraries for Apache Arrow
 
   - name: pyarrow-tests
-    script: build-pyarrow.sh  # [unix]
+    script: build-pyarrow.sh   # [unix]
     script: build-pyarrow.bat  # [win]
     version: {{ version }}
     build:

--- a/recipe/meta.yaml
+++ b/recipe/meta.yaml
@@ -234,7 +234,7 @@ outputs:
         - {{ compiler("cuda") }}                 # [cuda_compiler_version != "None"]
         - python                                 # [build_platform != target_platform]
         - cross-python_{{ target_platform }}     # [build_platform != target_platform]
-        - cython                                 # [build_platform != target_platform]
+        - cython <3                              # [build_platform != target_platform]
         - numpy                                  # [build_platform != target_platform]
         - cmake
         - ninja
@@ -244,7 +244,7 @@ outputs:
         - libarrow ={{ version }}=*_{{ PKG_BUILDNUM }}_{{ build_ext }}
         - clangdev {{ llvm_version }}
         - llvmdev {{ llvm_version }}
-        - cython
+        - cython <3
         - numpy
         - python
         - setuptools
@@ -318,7 +318,7 @@ outputs:
         - {{ compiler("cuda") }}                 # [cuda_compiler_version != "None"]
         - python                                 # [build_platform != target_platform]
         - cross-python_{{ target_platform }}     # [build_platform != target_platform]
-        - cython                                 # [build_platform != target_platform]
+        - cython <3                              # [build_platform != target_platform]
         - numpy                                  # [build_platform != target_platform]
         - cmake
         - ninja
@@ -327,7 +327,7 @@ outputs:
         - {{ pin_subpackage('pyarrow', exact=True) }}
         - clangdev {{ llvm_version }}
         - llvmdev {{ llvm_version }}
-        - cython
+        - cython <3
         - numpy
         - python
         - setuptools
@@ -355,7 +355,7 @@ outputs:
         - backports.zoneinfo     # [py<39]
         - cffi
         - cloudpickle
-        - cython
+        - cython <3
         - fastparquet
         - fsspec
         - hypothesis

--- a/recipe/meta.yaml
+++ b/recipe/meta.yaml
@@ -338,7 +338,7 @@ outputs:
       run_constrained:
         - apache-arrow-proc =*={{ build_ext }}
 
-    {% if not (aarch64 or ppc64le) or py == 311 %}
+    {% if not (aarch64 or ppc64le) or py in (310, 311) %}
     # only run the full test suite for one python version when in emulation (each run takes ~45min);
     # there's essentially zero divergence in behaviour across python versions anyway, and otherwise
     # CUDA builds for aarch/ppc consistently run out of disk space on azure for some reason

